ddtrace/internal/writer.py

```diff
@@ -60,41 +67,140 @@ def __init__(
         shutdown_timeout=DEFAULT_TIMEOUT,
         sampler=None,
         priority_sampler=None,
+        processing_interval=1,
+        buffer_size=16 * 1000000,  # 16MB
+        max_payload_size=8 * 1000000,  # 8MB
+        timeout=2,
     ):
         super(AgentWriter, self).__init__(
-            interval=self.QUEUE_PROCESSING_INTERVAL, exit_timeout=shutdown_timeout, name=self.__class__.__name__
+            interval=processing_interval, exit_timeout=shutdown_timeout, name=self.__class__.__name__
         )
-        # DEV: provide a _temporary_ solution to allow users to specify a custom max
-        maxsize = int(os.getenv("DD_TRACE_MAX_TPS", self.QUEUE_MAX_TRACES_DEFAULT))
-        self._trace_queue = _queue.TraceQueue(maxsize=maxsize)
+        self._buffer_size = buffer_size
+        self._max_payload_size = max_payload_size
+        self._buffer = TraceBuffer(max_size=self._buffer_size, max_item_size=self._max_payload_size)
         self._sampler = sampler
         self._priority_sampler = priority_sampler
-        self._last_error_ts = 0
-        self.api = api.API(
-            hostname, port, uds_path=uds_path, https=https, priority_sampling=priority_sampler is not None
-        )
-        if hasattr(time, "thread_time"):
-            self._last_thread_time = time.thread_time()
+        self._hostname = hostname
```

Comment:
While I agree with moving the logic for writing traces out of `API`, the writer now has the connection details, which we are accessing elsewhere in the library. 

What do you think about keeping the specifics of the connections themselves in the `api` module and adding a function to the writer returns the path of a connection (`uds://...`) which we use in our test cases rather than accessing internals like `_hostname` or `_port`?

Reply:
I don't think the indirection of having an `api` module is necessary. It does feel kinda nasty that we're reaching into internal fields, but I think we really only do this in tests and in `tracer.configure` (I think). Also, I think that it makes sense for the connection details to be on the writer. Maybe they should be public fields instead though?

Longer term I'm thinking that we'll support many modular writers (like otel) and then the writers will be completely decoupled from the tracer instance.

Reply:
Don't forget we have multiple writers now. We have the log writer for lambda.

We might want to add a "sync" writer in the future.

If we keep the API simple, could be easy for people to extend it as well, like sending in OTLP, or doing custom batching, sending strategies for specific use cases, etc

Reply:
This PR de-decouples the writer and the "api", which makes sense to me. I'm not sure about coupling the writer with the connection, i.e. the writer is concerned with building payloads from traces and putting those payloads onto a connection but a factory for such connections could be factored out. However, we can revisit this once we further decouple the writer from the tracer.

Reply:
> If we keep the API simple, could be easy for people to extend it as well, like sending in OTLP, or doing custom batching, sending strategies for specific use cases, etc

Yep, we're almost there now, it'd just be a matter of having the tracer be configured with a list of writers instead of a single one. We'd have to decide on a couple of things (eg. whether writers receive spans as they finish or if they receive lists of completed spans orrrr events!). Topic for another time though.

> I'm not sure about coupling the writer with the connection, i.e. the writer is concerned with building payloads from traces and putting those payloads onto a connection but a factory for such connections could be factored out. However, we can revisit this once we further decouple the writer from the tracer.

Ah gotcha. Yes we could, but I don't think the logic is sophisticated enough to factor out. With all the connection and encoding logic it's < 200 lines and is (imo) quite easy to grok. And like you said, we can revisit if it becomes an issue later on. 🙂